### PR TITLE
Support typedef'd (anonymous) structs ##types

### PR DESCRIPTION
**Pull request by kazarmy**
Partial fix for #12272.

### CODE REVIEW

**Comment by kazarmy** [comment]
Partial fix because typedefs and struct tags live in different namespaces. Shown by:

```c
struct A {
  float num;
};

typedef struct {
  double num;
} A;

int main(void) {
  printf("%zu\n", sizeof(struct A));
  printf("%zu\n", sizeof(A));
}
```
Output:
```
4
8
```

File: libr/core/cmd_type.c [hunk 1]
```diff
@@ -267,8 +267,22 @@ static void save_parsed_type(RCore *core, const char *parsed) {
 	}
 }
 
+static RCore *core_ = NULL; // HACK
+
 static int stdifstruct(void *user, const char *k, const char *v) {
-	return !strncmp (v, "struct", strlen ("struct") + 1);
+	if (!strncmp (v, "struct", strlen ("struct") + 1)) {
+		return true;
+	} else if (!strcmp (v, "typedef")) {
```
Comment (radare) [inline libr/core/cmd_type.c]: no need to else after return

File: libr/core/cmd_type.c [hunk 2]
```diff
@@ -267,8 +267,22 @@ static void save_parsed_type(RCore *core, const char *parsed) {
...
+		r_return_val_if_fail (core_ && core_->anal, false);
```
Comment (radare) [inline libr/core/cmd_type.c]: those statements should be at the begining of the function not here

File: libr/core/cmd_type.c [hunk 3]
```diff
@@ -267,8 +267,22 @@ static void save_parsed_type(RCore *core, const char *parsed) {
 	}
 }
 
+static RCore *core_ = NULL; // HACK
```
Comment (radare) [inline libr/core/cmd_type.c]: globals are evil

File: libr/core/cmd_type.c [hunk 4]
```diff
@@ -267,8 +267,22 @@ static void save_parsed_type(RCore *core, const char *parsed) {
...
+
 static int stdifstruct(void *user, const char *k, const char *v) {
-	return !strncmp (v, "struct", strlen ("struct") + 1);
+	if (!strncmp (v, "struct", strlen ("struct") + 1)) {
+		return true;
+	} else if (!strcmp (v, "typedef")) {
+		r_return_val_if_fail (core_ && core_->anal, false);
+		Sdb *TDB = core_->anal->sdb_types;
+		char key[136];
```
Comment (radare) [inline libr/core/cmd_type.c]: use sdb_fmt

File: libr/util/ctype.c [hunk 5]
```diff
@@ -423,6 +423,13 @@ R_API char *r_type_format(Sdb *TDB, const char *t) {
 		fmt = r_str_append (fmt, vars);
 		free (vars);
 		return fmt;
+	} else if (!strcmp (kind, "typedef")) {
+		snprintf (var2, sizeof (var2), "typedef.%s", t);
+		const char *type = sdb_const_get (TDB, var2, NULL);
+		// only supports struct atm
+		if (type && !strncmp (type, "struct ", 7)) {
+			return r_type_format (TDB, type + 7);
```
Comment (radare) [inline libr/util/ctype.c]: isnt type +7 the same as just t?

**Comment by XVilka** [comment]
I think it is a good approach because proper fix would require refactoring TCC code a lot. Good idea!

File: libr/core/cmd_type.c (hunk 3 above)
Comment (XVilka) [reply libr/core/cmd_type.c]: I don't like this part too.

**Comment by radare** [comment]
In fact you only need the TDB. Not core at all

File: libr/util/ctype.c (hunk 5 above)
Comment (kazarmy) [reply libr/util/ctype.c]: only for a specific case that should be rare but it causes infinite recursion so fixed in cd4f7164169f640f79b066eec0994379af4aad88

**Comment by XVilka** [comment]
Yes, problem is that typedefs and structures are totally different entities and we don't resolve typedefs when converting types into SDB representations.

File: libr/core/cmd_type.c (hunk 3 above)
Comment (kazarmy) [reply libr/core/cmd_type.c]: I don't like it either but doing away with the global apparently requires significant work in sdb -- there should be a way to pass a context to sdb_foreach_list_filter().

**Comment by kazarmy** [comment]
> In fact you only need the TDB. Not core at all

Done in 055c05b56851f649eeb50219e534cf1404fcbff5.

> ... we don't resolve typedefs when converting types into SDB representations.

I don't think we should.

File: libr/core/cmd_type.c (hunk 1 above)
Comment (kazarmy) [reply libr/core/cmd_type.c]: done 2a9a424fa6527c0b43a725c75f6b26bd0be3f848

File: libr/core/cmd_type.c (hunk 2 above)
Comment (kazarmy) [reply libr/core/cmd_type.c]: done 466a38fc1987c892091597bff34a65ab5e60d6a2

File: libr/core/cmd_type.c (hunk 4 above)
Comment (kazarmy) [reply libr/core/cmd_type.c]: done 4f5b655e2fcd13cd14409f8acb4773c67b64ec7b

File: libr/core/cmd_type.c (hunk 3 above)
Comment (kazarmy) [reply libr/core/cmd_type.c]: That should be in a different PR imho.

File: libr/core/cmd_type.c [hunk 6]
```diff
@@ -267,8 +267,21 @@ static void save_parsed_type(RCore *core, const char *parsed) {
 	}
 }
 
+static Sdb *TDB_ = NULL; // HACK
+
 static int stdifstruct(void *user, const char *k, const char *v) {
-	return !strncmp (v, "struct", strlen ("struct") + 1);
+	r_return_val_if_fail (TDB_, false);
+	if (!strncmp (v, "struct", strlen ("struct") + 1)) {
```
Comment (radare) [inline libr/core/cmd_type.c]: uhm, isnt this the same as !strcmp ? because you are comparing the trailing 0 too

File: libr/core/cmd_type.c [hunk 7]
```diff
@@ -267,8 +267,21 @@ static void save_parsed_type(RCore *core, const char *parsed) {
...
+		return true;
+	}
+	if (!strcmp (v, "typedef")) {
+		const char *typedef_key = sdb_fmt ("typedef.%s", k);
+		const char *type = sdb_const_get (TDB_, typedef_key, NULL);
+		if (type && !strncmp (type, "struct ", 7)) {
```
Comment (radare) [inline libr/core/cmd_type.c]: for readability we can use r_str_startswith()

File: libr/core/cmd_type.c (hunk 6 above)
Comment (kazarmy) [reply libr/core/cmd_type.c]: yes ... fixed old code c8c0efe31672fbed619a054654d368f12da668eb

File: libr/core/cmd_type.c (hunk 7 above)
Comment (kazarmy) [reply libr/core/cmd_type.c]: done 6c632e94cc0bc88a603a036214ed81aa2b431deb

**Comment by kazarmy** [comment]
Just be aware that usage of a counter for anon structs can cause problems, e.g. #13014, and with preexisting SDB entries.

**Pull request merged**